Apply dynamic_cast consistently

ddn0 opened the pull request:
```
Break the use of dynamic_cast into cases and apply them consistently.

If the cast is assumed to succeed, use either

  auto& ret = dynamic_cast<T&>(item)

or

  auto* ret = dynamic_cast<T*>(item); KATANA_LOG_ASSERT(ret)

In both cases, the program will crash if the assumption does not hold.
This is preferable to memory corruption.

If the cast could fail, use

  auto* ret = dynamic_cast<T*>(item)

And if the cast involves a shared_ptr, use

  auto ret = std::dynamic_pointer_cast<T>(item)

unless the lifetime of item can be guaranteed and the situation requires
extreme performance tuning. dynamic_pointer_cast makes sure that the
lifetime of item and ret are linked and avoids dangling pointers. This
is preferable to memory corruption or a slight performance penalty.
```

ddn0 commented on libquery/include/querying/Errors.h:
```diff
@@ -0,0 +1,77 @@
+
+#include "galois/config.h"
+
+namespace Querying {
```
Let's just stick to one namespace, `galois`. If you need a separate error code enum besides galois::ErrorCode, why not just create `galois::QueryErrorCode`?

ddn0 commented on libquery/src/OrderByOp.cpp:
```diff
@@ -63,27 +67,31 @@ galois::OrderByOp::Compute() {
     }
     default:
-      GALOIS_LOG_FATAL("Type {} is not supported", col.GetType());
+      GALOIS_LOG_ERROR("Type {} is not supported", col.GetType());
```
in other code, we have been using `GALOIS_LOG_DEBUG` for the extra details about errors

roshandathathri commented on libquery/src/CypherCompiler.cpp:
```diff
@@ -964,11 +1084,14 @@ galois::CypherCompiler::CompileASTNode(CypherASTNode ast) {
+      auto res = CompileASTNode(pattern);
+      if (!res) {
+        return res.error();
+      } else {
```
```
}
return ResultSuccess();
```

roshandathathri reviewed (approved):
This is great!

ddn0 commented on libquery/src/CypherCompiler.cpp:
```diff
@@ -372,39 +385,74 @@ galois::CypherCompiler::CompilePatternPath(CypherASTNode ast, bool negate) {
-          CompileNodePatternPath(first, this->query_edges.back().caused_by);
-          CompileRelPatternPath(rel);
-          CompileNodePatternPath(second, this->query_edges.back().acted_on);
+          auto res =
```
you can write these more compactly as:
```cpp
if (auto res = ...; !res) {
  return res.error();
}
```

ddn0 commented on libquery/src/CypherCompiler.cpp:
```diff
@@ -1067,7 +1189,11 @@ galois::CypherCompiler::Compile(const char* queryStr) {
   if (nerrors == 0) {
     // take ast, change it to a query graph
-    this->CompileAST(result);
+    auto res = CompileAST(result);
```
:)

danielmawhirter commented on libquery/src/LimitOp.cpp:
```diff
@@ -2,7 +2,7 @@
 uint32_t galois::LimitOp::num_limit_ops_ = 0;
 
-std::vector<galois::Operator*>
+galois::Result<std::vector<galois::Operator*>>
```
No control path returns a failure here

danielmawhirter commented on libquery/src/SkipOp.cpp:
```diff
@@ -2,7 +2,7 @@
 uint32_t galois::SkipOp::num_skip_ops_ = 0;
 
-std::vector<galois::Operator*>
+galois::Result<std::vector<galois::Operator*>>
```
ditto

danielmawhirter commented on libquery/src/CountFunc.cpp:
```diff
@@ -5,7 +5,7 @@
 #include "galois/gIO.h"
 
-galois::Column
+galois::Result<galois::Column>
```
ditto

danielmawhirter reviewed (approved):
LGTM, some nits

roshandathathri commented on libquery/include/querying/CypherCompiler.h:
```diff
@@ -200,6 +202,10 @@ class GALOIS_EXPORT CypherCompiler {
+  // TODO (bwu): This function does not return
+  // Result<std::unique_ptr<CompilerOrderByMetadata>> because gcc-7 failed to convert
+  // std::unique_ptr<CompilerOrderByMetadata> to
+  // Result<std::unique_ptr<CompilerOrderByMetadata> while gcc-9 succeeded.
```
Is there a way to explicitly convert the type? @ddn0

ddn0 replied:
Yeah:
https://github.com/KatanaGraph/katana/blob/master/libsupport/include/galois/Result.h#L25

danielmawhirter replied:
Adopting that in #556

ddn0 replied:
Maybe we should put this hint in `CONTRIBUTING.md`?

bowu replied:
That's a good idea.

ddn0 committed:
Apply dynamic_cast consistently

Break the use of dynamic_cast into cases and apply them consistently.

If the cast is assumed to succeed, use either

  auto& ret = dynamic_cast<T&>(item)

or

  auto* ret = dynamic_cast<T*>(item); KATANA_LOG_ASSERT(ret)

In both cases, the program will crash if the assumption does not hold.
This is preferable to memory corruption.

If the cast could fail, use

  auto* ret = dynamic_cast<T*>(item)

And if the cast involves a shared_ptr, use

  auto ret = std::dynamic_pointer_cast<T>(item)

unless the lifetime of item can be guaranteed and the situation requires
extreme performance tuning. dynamic_pointer_cast makes sure that the
lifetime of item and ret are linked and avoids dangling pointers. This
is preferable to memory corruption or a slight performance penalty.

rob-isaac reviewed (approved):
Looks good
